modules/custom/az_paragraphs/src/AZBackgroundImageCssHelper.php

```diff
@@ -0,0 +1,150 @@
+        'bg_image_repeat' => 'no-repeat',
+        'bg_image_background_size' => '',
+        'bg_image_background_size_ie8' => 0,
+        'bg_image_gradient' => '',
+        'bg_image_media_query' => 'all',
+        'bg_image_important' => 1,
+        'bg_image_z_index' => 'auto',
+        'bg_image_path_format' => 'absolute',
+      ],
+    ];
+  }
+
+  /**
+   * Function taken from the module 'bg_image'.
+   *
+   * Adds a background image to the page using the
+   * css 'background' property.
+   *
+   * @param string $image_path
+   *   The path of the image to use. This can be either
+   *      - A relative path e.g. sites/default/files/image.png
+   *      - A uri: e.g. public://image.png.
+   * @param array $css_settings
+   *   An array of css settings to use. Possible values are:
+   *      - bg_image_selector: The css selector to use
+   *      - bg_image_color: The background color
+   *      - bg_image_x: The x offset
+   *      - bg_image_y: The y offset
+   *      - bg_image_attachment: The attachment property (scroll or fixed)
+   *      - bg_image_repeat: The repeat settings
+   *      - bg_image_background_size: The background size property if necessary
+   *    Default settings will be used for any values not provided.
+   * @param string $image_style
+   *   Optionally add an image style to the image before applying it to the
+   *   background.
+   *
+   * @return array
+   *   The array containing the CSS.
+   */
+  public function getBackgroundImageCss($image_path, array $css_settings = [], $image_style = NULL) {
+    $defaults = self::defaultSettings();
+    $css_settings += $defaults['css_settings'];
+
+    // Pull the default css setting if not provided.
+    $selector = Xss::filter($css_settings['bg_image_selector']);
+    $bg_color = Xss::filter($css_settings['bg_image_color']);
+    $bg_x = Xss::filter($css_settings['bg_image_x']);
+    $bg_y = Xss::filter($css_settings['bg_image_y']);
+    $attachment = $css_settings['bg_image_attachment'];
+    $repeat = $css_settings['bg_image_repeat'];
+    $important = $css_settings['bg_image_important'];
+    $background_size = Xss::filter($css_settings['bg_image_background_size']);
+    $background_size_ie8 = $css_settings['bg_image_background_size_ie8'];
+    $background_gradient = !empty($css_settings['bg_image_gradient']) ? $css_settings['bg_image_gradient'] . ',' : '';
+    $media_query = isset($css_settings['bg_image_media_query']) ? Xss::filter($css_settings['bg_image_media_query']) : NULL;
+    $z_index = Xss::filter($css_settings['bg_image_z_index']);
+
+    // If important is true, we turn it into a string for css output.
+    if ($important) {
+      $important = '!important';
+    }
+    else {
+      $important = '';
+    }
+
+    // Handle the background size property.
+    $bg_size = '';
+    $ie_bg_size = '';
+
+    if ($background_size) {
+      // CSS3.
+      $bg_size = sprintf('background-size: %s %s;', $background_size, $important);
```

Comment:
I think it might be preferable to use [`FormattableMarkup::placeholderFormat()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Render%21FormattableMarkup.php/function/FormattableMarkup%3A%3AplaceholderFormat/9.2.x) instead of `sprintf()`.

Reply:
FYI FormattableMarkup says not to use for css.
https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Render%21FormattableMarkup.php/class/FormattableMarkup/9.3.x

Reply:
I think that only applies to one of the placeholder types.

Reply:
Drupal core issue Re: `sprintf()`
https://www.drupal.org/project/drupal/issues/3118957